tls/s2n_fingerprint_ja4.c

```diff
@@ -0,0 +1,542 @@
+    RESULT_GUARD(s2n_fingerprint_hash_add_bytes(hash, ianas, iana_list_size - 1));
+    return S2N_RESULT_OK;
+}
+
+static S2N_RESULT s2n_fingerprint_ja4_sig_algs(struct s2n_fingerprint_hash *hash,
+        struct s2n_client_hello *ch)
+{
+    s2n_parsed_extension *extension = NULL;
+    RESULT_GUARD_POSIX(s2n_client_hello_get_parsed_extension(
+            S2N_EXTENSION_SIGNATURE_ALGORITHMS, &ch->extensions, &extension));
+
+    struct s2n_stuffer sig_algs = { 0 };
+    RESULT_GUARD_POSIX(s2n_stuffer_init_written(&sig_algs, &extension->extension));
+
+    uint8_t entry_bytes[S2N_JA4_IANA_ENTRY_SIZE] = { 0 };
+    struct s2n_stuffer entry = { 0 };
+    RESULT_GUARD_POSIX(s2n_blob_init(&entry.blob, entry_bytes, sizeof(entry_bytes)));
+
+    bool is_first = true;
+    RESULT_GUARD_POSIX(s2n_stuffer_skip_read(&sig_algs, sizeof(uint16_t)));
+    while (s2n_stuffer_data_available(&sig_algs)) {
+        uint16_t iana = 0;
+        RESULT_GUARD_POSIX(s2n_stuffer_read_uint16(&sig_algs, &iana));
+        if (s2n_is_grease_value(iana)) {
+            continue;
+        }
+        if (!is_first) {
+            RESULT_GUARD_POSIX(s2n_stuffer_write_char(&entry, S2N_JA4_LIST_DIV));
+        }
+        RESULT_GUARD(s2n_stuffer_write_uint16_hex(&entry, iana));
+        RESULT_GUARD(s2n_fingerprint_hash_add_bytes(hash, entry_bytes,
+                s2n_stuffer_data_available(&entry)));
+        RESULT_GUARD_POSIX(s2n_stuffer_rewrite(&entry));
+        is_first = false;
+    }
+    return S2N_RESULT_OK;
+}
+
+/**
+ *= https://raw.githubusercontent.com/FoxIO-LLC/ja4/v0.18.2/technical_details/JA4.md#extension-hash
+ *# A 12 character truncated sha256 hash of the list of extensions, sorted by
+ *# hex value, followed by the list of signature algorithms, in the order that
+ *# they appear (not sorted).
+ */
+static S2N_RESULT s2n_fingerprint_ja4_c(struct s2n_fingerprint *fingerprint,
+        struct s2n_fingerprint_hash *hash, struct s2n_stuffer *output,
+        struct s2n_blob *extensions_count)
+{
+    uint16_t extensions_count_value = 0;
+    RESULT_GUARD(s2n_fingerprint_ja4_extensions(hash, fingerprint->client_hello,
+            &fingerprint->workspace, &extensions_count_value));
+
+    /**
+     *= https://raw.githubusercontent.com/FoxIO-LLC/ja4/v0.18.2/technical_details/JA4.md#extension-hash
+     *# The signature algorithm hex values are then added to the end of the list
+     *# in the order that they appear (not sorted) with an underscore delimiting
+     *# the two lists.
+     *
+     *= https://raw.githubusercontent.com/FoxIO-LLC/ja4/v0.18.2/technical_details/JA4.md#extension-hash
+     *# If there are no signature algorithms in the hello packet,
+     *# then the string ends without an underscore and is hashed.
+     */
+    RESULT_GUARD(s2n_fingerprint_hash_add_char(hash, S2N_JA4_PART_DIV));
+    /* Ignore missing or malformed signature algorithm */
+    s2n_result_ignore(s2n_fingerprint_ja4_sig_algs(hash, fingerprint->client_hello));
```

Comment:
Shouldn't the underscore not be added if there are no signature algorithms?

Reply:
You have no idea how many times I read the spec as "then the string ends WITH an underscore" instead of "WITHOUT an underscore" D: I even wrote a test for it! How?!

Thank you for catching that lol